### ISSUE TITLE
Thin lines not rendered correctly

### ISSUE THREAD

**Issue by allkhor** [godotengine/godot#24207]
**Godot version:**

Godot 3.1 master 21dea10, Godot 2.1.5

**OS/device including version:**

Xubuntu 16.04.5 LTS, Nvidia GTX 750 Ti, proprietary driver ver. 390.87

**Issue description:**

When used 2d/viewport mode thin lines not rendered correctly.
It's affected LineEdit caret.
![thin_lines](https://user-images.githubusercontent.com/1049442/49637616-14d9ea80-fa30-11e8-96de-2ac7a390a760.gif)

**Steps to reproduce:**
1) Open attached project.
2) Run scene.
3) Some lines invisible.

**Minimal reproduction project:**

For Godot 3.1: [thin_lines.zip](https://github.com/godotengine/godot/files/2656533/thin_lines.zip)

**Comment by CptPotato** [comment]
This heavily depends on the line's positioning, scale aswell as the scale of the viewport.

If they don't align to the pixels correctly, missing lines or ones that are too thick are the result.
This is the standard behavior and I think this is okay.

What may happen here is that the container adds a small offset to the node causing a subpixel offset. Have you tried enabling `pixel snap`™?

**Comment by allkhor** [comment]
Use Pixel Snap (Rendering/Quality/2d) don't solve problem.
I expected lines work like a guides lines in Godot Editor or stroke in viewport, that's visible in any situation.

**Comment by CptPotato** [comment]
Lines in the editor are drawn differently. They don't use textures but OpenGL functionality to do this.
I'm not sure if Godot has support for actual line meshes at the moment.

**Comment by guilhermefelipecgs** [comment]
> I expected lines work like a guides lines in Godot Editor or stroke in viewport, that's visible in any situation.

You should use `draw_line` then.

**Comment by allkhor** [comment]
My game used LineEdit, how to fix invisible caret in some position?
Increase caret width, sounds like a workaround.
https://github.com/godotengine/godot/blob/19f2a188bb305c4a446f63a0b5417086b7261514/scene/gui/line_edit.cpp#L823

**Comment by guilhermefelipecgs** [comment]
> My game used LineEdit, how to fix invisible caret in some position?
> Increase caret width, sounds like a workaround.

This problems occurs because you are using  `display/window/size/test_widht` and `test_height` with low resolution, changing to zero will fix that.

**Comment by allkhor** [comment]
It's set for test purpose, changing to zero not fixed bug, if user change windows size?

**Comment by guilhermefelipecgs** [comment]
yeah, if windows is resizable than caret can disappear indeed. But in that case I think this is expected because you are using stretch mode. The only way of fixing it it's disabling stretch.

**Comment by guilhermefelipecgs** [comment]
For example, if I resize my browser, the interface like search field, tabs, buttons, fonts, etc, doesn't change their sizes, only the content of the site if it's responsible design.

**Comment by guilhermefelipecgs** [comment]
I think this may be a reason why some games don't allow resizing the window, usually they give some pre-defined resolutions for the player to choose. I personally would go that way.

**Comment by AlexHolly** [comment]
https://github.com/godotengine/godot/commit/45e7306b5adec09746fefda902a5414e5b53fdfa

Having a similar issue, shows different lines on move, resize and scale.
`pixel snap` has no effect

![peek 2019-03-05 15-08](https://user-images.githubusercontent.com/4741886/53811020-a33f0900-3f58-11e9-9807-99dae4fabe26.gif)

[LineRender.zip](https://github.com/godotengine/godot/files/2931256/LineRender.zip)

**Comment by KoBeWi** [comment]
Still valid in 481151b
